### ISSUE TITLE
WebGPU support

### ISSUE THREAD

**Issue by ca1773130n** [aras-p/UnityGaussianSplatting#65]
Thanks for sharing your great work.
I've been investigating about how to deploy for web browsers.

WebGL doesn't work as there is no compute shaders, so we need WebGPU backend.
Fortunately, Unity 2023 beta supports experimental WebGPU though some scripting for editor is needed to enable it.

For now I'm struggling with radix sort codes from FidelityFX, to make it work on WebGPU.
Here's some issues:

**It seems like WGSL does not have a counterpart of GroupMemoryBarrierWithGroupSync().**
- It's being converted to just workgroupBarrier() which will slow down the sorting algorithm.
- Also the shader compilation fails due to calling the function on a non-uniform control bound.
- I couldn't succeeded to modify the original HLSL code to make it work.

**wavebasic does not work. it produces an error like: invalid kernel index passed.**
- Metal backend works without it, so I can just comment it out but it could be much slower

There are working WebGPU implementations for 3D GS, so hopefully we could modify the current code to make it run without errors.

**Comment by aras-p** [comment]
Yeah all the shaders there right now are very much geared towards "let's assume we have a full modern shader capability", which WebGPU is still lacking in some aspects. I've spent zero effort to get this working on mobile or the web. Doable, but was not my focus.

**Comment by munrocket** [comment]
Are you sure that you need GroupMemoryBarrierWithGroupSync()? Seems DX doesn’t support it and all executed in lock-step.

**Comment by maki0315** [comment]
I reported this error after Posting :

**Comment by maki0315** [comment]
![image](https://github.com/aras-p/UnityGaussianSplatting/assets/62171024/43c75b7b-89b1-416c-a8c1-41a95768430b)

**Comment by lucylucy27** [comment]
Hi， [ca1773130n](https://github.com/ca1773130n)
Did you later successfully complete the support work for WebGPU? 
Can this project correctly open and view PLY files in Chrome?
Thanks ~

**Comment by crazyzyz** [comment]
Hello, I want to try to use webgl package to generate in chrome now, but after chrome runs, there is a black screen. Through my check of browser developer tools, I find that there is no compute shader. Secondly, I find that the lack of compute shader is due to the lack of browser. I look forward to your reply, which is very important to me

**Comment by aras-p** [comment]
@crazyzyz WebGPU is not supported right now. This very issue is open for a reason!

**Comment by crazyzyz** [comment]
> 目前不支持 WebGPU。这个问题是有原因的！

Do you know why

**Comment by aras-p** [comment]
> Do you know why

Because I have not implemented the right things needed for WebGPU support. The largest issue currently is probably lack of GPU sorting, since WebGPU (right now) does not have "wave/subgroup" shader operations.

**Comment by Zarbuz** [comment]
WebGPU subgroup operations will be enabled by default in Chrome 134 : https://chromestatus.com/feature/5126409856221184

**Comment by FancyB1ts** [comment]
@ca1773130n @andybak @crazyzyz @Zarbuz Has anyone of you been looking into this topic recently, now that Chrome supports WebGPU subgroup operations?

**Comment by Zarbuz** [comment]
@FancyB1ts Now we have to wait for the implementation on the Unity side (I continue to follow the developments of WebGPU on the Unity side with great interest).  The limitations are still relevant in Unity 6.2 . Maybe with 6.3...

**Comment by brendan-duncan** [comment]
Adding to this thread since I was just asked about it. 

WebGPU subgroup operations require DXC shader compiler, which is still a work-in-progress for WebGPU shaders in Unity. A long while back I had forked this repo to use a different sorting method for WebGPU that doesn't require the subgroup operators, https://github.com/brendan-duncan/UnityGaussianSplatting. I never got around to making a PR for Aras, and I see that Aras has made some changes since I had done that fork. I'll try to get my fork updated in the near future, and see if Aras wants the changes via a PR so that WebGPU will work until subgroup support can be added via the DXC to the Unity shader compiler for WebGPU.

**Comment by aras-p** [comment]
@brendan-duncan hmm yeah, originally this whole project was supposed to be just a tiny playground for me to learn about 3DGS, as well as to learn about GPU sorting. As such, I had no interest whatsoever in having GPU sorting code paths that would "work everywhere"; and much more interested in "what would be state of the art" aspect.

Right now I am not even working on this project all that much (at all?) anymore, so from my POV it would be easiest to just wait for DXC being used for WebGPU target :)

**Comment by brendan-duncan** [comment]
I have plenty of my own side projects that continue to haunt me :-) 
People can use the forks if they want WebGPU.
